Add an optional additional argument for visualize_{} functions.

grigorisg9gr opened the pull request:
This optional argument prints a message (element) for each image.

Currently, there is no way to print some additional info in the info box. 
This can be useful if you want to print additional info along with the images, e.g. the angle of the image, the fitting error of the saved landmarks, any non-included message. 

New code that works for me (apart from the old options that work fine):

```
import menpo.io as mio
c = mio.import_builtin_asset
ims = [c.lenna_png(), c.takeo_ppm(), c.einstein_jpg()]
try:
    %matplotlib inline
    from menpowidgets import visualize_images
except NameError:
    pass

visualize_images(ims, messages=list(range(3)))
```

grigorisg9gr committed:
Add an optional additional argument for visualize_images().

This optional argument prints a message (element) for each image.

patricksnape commented:
If you really want this, I feel like it would be better as a callback function? So you pass a function and and you get given each image every time you load it and you can do whatever you want and just return a string? @jabooth @nontas thoughts?

nontas commented:
I don't really see why this is useful to be honest. But anyway in case we agree to add it, then the following should be done:
1. The default value of `messages` should not be an empty str, but `None`.
2. The same argument should be added to all our widgets (`visualize_pointclouds`, `visualize_landmarkgroups` etc.).

nontas commented on menpowidgets/base.py:
```diff
@@ -650,7 +650,7 @@ def update_widgets(change):
 
 def visualize_images(images, figure_size=(10, 8), style='coloured',
-                     browser_style='buttons'):
+                     browser_style='buttons', messages=''):
```
It's best to make this `messages=None`

nontas commented on menpowidgets/base.py:
```diff
@@ -673,6 +673,9 @@ def visualize_images(images, figure_size=(10, 8), style='coloured',
     browser_style : ``{'buttons', 'slider'}``, optional
         It defines whether the selector of the objects will have the form of
         plus/minus buttons or a slider.
+    messages: `list` with the same length as the `images`, optional
```
This should be

```
messages : `list` or ``None``, optional
```

And then explain that it must have the same length as `images` in the description below.

patricksnape commented:
@nontas Well that's why I was suggesting a callback:

```
def custom_print(obj):
    return 'Some custom message: {}'.format(object.some_attribute)

visualize_images(images, custom_info_callback=custom_print)
```

So at least you could print anything you like and have it automatically calculated per image. Would be more useful for usages with LazyLists.

grigorisg9gr commented:
Patrick,  why are you suggesting to pass a function rather than a (lazy) list as an argument? 

I can make the changes you propose @nontas , sure, but would you prefer that to Patrick's version?

patricksnape commented:
If you have a lazy list of images and you want to print something about each image that gets computed, like the average pixel intensity or something, you don't want to load every image to compute it and pass it is a list of strings. Nor do you want to load the image twice which is what you would have to do to create a new lazy list that calculates the value. You want a function that gets passed the image as you cycle through the visualize method.

nontas commented:
Yeah, if we are going to add such a feature, then I agree with @patricksnape approach. The callback could even return a `list` of `str`, where the length of the list defines the number of additional lines that will be appended at the Info tab. Of course, this has to be defined for all widget functions.

grigorisg9gr committed:
Implement the custom message printing as a callback function.

grigorisg9gr commented:
I implemented it (i believe) in the way you described @patricksnape after discussing with @nontas . Please check it out, if you are happy with it, I can make the changes in the rest widgets.

nontas commented on menpowidgets/base.py:
```diff
@@ -650,7 +650,7 @@ def update_widgets(change):
 
 def visualize_images(images, figure_size=(10, 8), style='coloured',
-                     browser_style='buttons'):
+                     browser_style='buttons', custom_function=None):
```
I think that the `custom_info_callback` variable name initially proposed by @patricksnape is more clear than `custom_function`.

nontas commented on menpowidgets/base.py:
```diff
@@ -673,6 +673,9 @@ def visualize_images(images, figure_size=(10, 8), style='coloured',
         It defines whether the selector of the objects will have the form of
         plus/minus buttons or a slider.
+    custom_function: `function` or ``None``, optional
+        If not None, it should be a function that accepts an image and returns
```
Maybe it should return a `list` of `str`, where each `str` is an additional line in the info tab?

nontas commented on menpowidgets/base.py:
```diff
@@ -793,6 +800,8 @@ def update_info(img, image_is_masked, group):
             text_per_line.append("> {} landmark points".format(
                 img.landmarks[group].lms.n_points))
             n_lines += 1
+        if custom_str is not None:
```
You're actually checking if the provided custom function argument is `None` twice: before calling the `update_info` function (line 769) and inside the function (line 803). I think you should pass in the custom function directly as an argument in `update_info()`. Then the check will only happen once in the function and in case it is not `None` you can call it as `custom_info_callback(img)`.

nontas commented on menpowidgets/base.py:
```diff
@@ -793,6 +800,8 @@ def update_info(img, image_is_masked, group):
                 img.landmarks[group].lms.n_points))
             n_lines += 1
+        if custom_str is not None:
+            text_per_line.append('> custom message: {}'.format(custom_str))
```
Don't print the `'custom message:'` string, especially if the custom function will return multiple lines. The user can define the custom function to return any string they want. However, do keep the `'> '` in front of each new line that is added in the info tab.

grigorisg9gr committed:
Allow the custom function to return a list of strings.

patricksnape commented on menpowidgets/base.py:
```diff
@@ -766,17 +767,14 @@ def render_function(change):
 
     # Define function that updates the info text
-    def update_info(img, image_is_masked, group, custom_str=None):
+    def update_info(img, image_is_masked, group, custom_info_callback):
```
@grigorisg9gr Tiny nitpick. If something is expected to _maybe_ be `None`, we generally use keyword arguments. This lets the developer know that the argument is optional. So, although its a touch verbose, it would be good if this was `custom_info_callback=None` and the above call is changed to `custom_info_callback=custom_info_callback`

patricksnape commented:
Once that is made a kwarg i will merge it in. Nice job @grigorisg9gr.

grigorisg9gr committed:
Update all the visualizations with the optional callback function.

grigorisg9gr committed:
Delete the n_lines from visualizations in base.py.

n_lines was not used in any computation, probably redundant from a previous update.

nontas commented:
Good job @grigorisg9gr 
+1
